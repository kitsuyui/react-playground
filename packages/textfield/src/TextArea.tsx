import {
  ComponentPropsWithoutRef,
  forwardRef,
  useCallback,
  useRef,
  useState,
} from 'react'
import React from 'react'

import { useCombinedRefs } from './utils'

type WrappedProps = ComponentPropsWithoutRef<'textarea'>

type alternateProps = {
  onInputChunk?: (value: string) => void
  onChangeInputting?: (inputting: boolean) => void
  value?: string
}

type excludeProps =
  | 'onChange'
  | 'onCompositionStart'
  | 'onCompositionUpdate'
  | 'onCompositionEnd'
  | 'value'

type WrapperProps = Omit<WrappedProps, excludeProps> & alternateProps

export const TextArea = forwardRef<HTMLTextAreaElement, WrapperProps>(
  (props, ref) => {
    const { onInputChunk, onChangeInputting } = props
    const [internalValue, setInternalValue] = useState(props.value ?? '')
    const [isInputting, setIsInputting] = useState(false)
    const innerRef = useRef<HTMLTextAreaElement>(null!)
    const compbinedRef = useCombinedRefs(innerRef, ref)

    // Use Object.assign({}, props) instead of { ...props } because it must create deep copy.
    const propsExcludedWrapperProps = Object.assign({}, props)
    delete propsExcludedWrapperProps.onInputChunk
    delete propsExcludedWrapperProps.onChangeInputting

    const handleChange = useCallback(() => {
      const text = innerRef.current.value
      setInternalValue(text)
      if (isInputting) return
      onInputChunk?.(text)
    }, [isInputting, onInputChunk])

    const handleCompositionChange = useCallback(() => {
      const text = innerRef.current.value
      setInternalValue(text)
      setIsInputting(true)
      onChangeInputting?.(true)
    }, [innerRef, onChangeInputting])

    const handleCompositionEnd = useCallback(() => {
      const text = innerRef.current.value
      setInternalValue(text)
      setIsInputting(false)
      onChangeInputting?.(false)
      onInputChunk?.(text)
    }, [innerRef, onInputChunk, onChangeInputting])

    const handleOnBlur = useCallback(
      (e: React.FocusEvent<HTMLTextAreaElement>) => {
        const text = innerRef.current.value
        setInternalValue(text)
        setIsInputting(false)
        onChangeInputting?.(false)
        onInputChunk?.(text)
        propsExcludedWrapperProps.onBlur?.(e)
      },
      [innerRef, onInputChunk, onChangeInputting, propsExcludedWrapperProps]
    )

    return (
      <textarea
        {...propsExcludedWrapperProps}
        ref={compbinedRef}
        value={internalValue}
        onCompositionStart={handleCompositionChange}
        onCompositionUpdate={handleCompositionChange}
        onCompositionEnd={handleCompositionEnd}
        onBlur={handleOnBlur}
        onChange={handleChange}
      />
    )
  }
)

TextArea.displayName = 'TextArea'
