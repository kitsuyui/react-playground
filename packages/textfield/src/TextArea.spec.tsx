import { expect, it, describe, vi, beforeEach } from 'vitest'

import { render, cleanup, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { TextArea, type TextAreaRef } from './TextArea'
import React from 'react'

const ExampleTextArea = (props: {
  initialText?: string
  onChange?: (value: string) => void
  onChangeInputting?: (inputting: boolean) => void
}) => {
  const { initialText, onChange, onChangeInputting } = props
  const ref = React.useRef<TextAreaRef>(null)
  const [text, setText] = React.useState(initialText ?? '')
  return (
    <>
    <TextArea
      value={text}
      ref={ref}
      onInputChunk={(chunk) => {
        setText(chunk)
        onChange?.(chunk)
      }}
      onChangeInputting={(inputting) => {
        onChangeInputting?.(inputting)
      }}
    />
    <button
      type="button"
      onClick={() => {
        ref.current?.clear()
      }}
    >
      Clear
    </button>
    </>
  )
}

beforeEach(() => {
  cleanup()
})

describe('TextArea', () => {
  it('should render', () => {
    const { asFragment } = render(
      <TextArea value="Hello, World!" onInputChunk={() => {}} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('treats empty string when value is not provided', () => {
    const { asFragment } = render(
      <TextArea/>
    )
    expect(asFragment()).toMatchSnapshot()
    const textarea = document.querySelector('textarea') as HTMLTextAreaElement
    expect(textarea).toBeInstanceOf(HTMLTextAreaElement)
    expect(textarea.value).toBe('')
  })

  it('should handle input', async () => {
    const handleInputChunk = vi.fn()
    const rendered = render(
      <ExampleTextArea
        initialText=""
        onChange={handleInputChunk}
      />
    )
    expect(rendered.asFragment()).toMatchSnapshot()
    const textarea = rendered.getByRole('textbox') as HTMLTextAreaElement
    expect(textarea).toBeInstanceOf(HTMLTextAreaElement)
    expect(textarea.value).toBe('')
    await userEvent.click(textarea) // focus the textarea
    await userEvent.type(textarea, 'Hello, World!')
    expect(textarea.value).toBe('Hello, World!')
    expect(rendered.asFragment()).toMatchSnapshot()
  })

  it('should handle input with composition change events', async () => {
    const handleInputChunk = vi.fn()
    const handleInputting = vi.fn()
    const rendered = render(
      <ExampleTextArea
        initialText=""
        onChange={handleInputChunk}
        onChangeInputting={handleInputting}
      />
    )
    expect(rendered.asFragment()).toMatchSnapshot()
    const textarea = rendered.getByRole('textbox') as HTMLTextAreaElement
    expect(textarea).toBeInstanceOf(HTMLTextAreaElement)
    expect(textarea.value).toBe('')
    await userEvent.click(textarea) // focus the textarea
    // userEvent.type does not support composition events directly,
    // so we simulate composition events manually
    expect(handleInputting).toHaveBeenCalledTimes(0)
    fireEvent.compositionStart(textarea)
    expect(handleInputting).toHaveBeenCalledTimes(1)
    fireEvent.change(textarea, { target: { value: 'Hello' } })
    expect(handleInputting).toHaveBeenCalledTimes(1)
    // When composition is in progress, textarea value is changed but not called onInputChunk
    expect(textarea.value).toBe('Hello')
    expect(handleInputChunk).toHaveBeenCalledTimes(0)

    // After composition ends, the value is committed
    fireEvent.compositionEnd(textarea)
    expect(handleInputting).toHaveBeenCalledTimes(2)
    expect(textarea.value).toBe('Hello')
    expect(handleInputChunk).toHaveBeenCalledTimes(1)
    expect(handleInputChunk).toHaveBeenCalledWith('Hello')
  })

  it('clear should reset value', async () => {
    const handleInputChunk = vi.fn()
    const rendered = render(
      <ExampleTextArea
        initialText="Hello, World!"
        onChange={handleInputChunk}
      />
    )
    expect(rendered.asFragment()).toMatchSnapshot()
    const textarea = rendered.getByRole('textbox') as HTMLTextAreaElement
    expect(textarea).toBeInstanceOf(HTMLTextAreaElement)
    expect(textarea.value).toBe('Hello, World!')
    await userEvent.click(textarea) // focus the textarea
    await userEvent.type(textarea, '2')
    expect(textarea.value).toBe('Hello, World!2')
    expect(rendered.asFragment()).toMatchSnapshot()
    const clearButton = rendered.queryByText('Clear')
    expect(clearButton).toBeInstanceOf(HTMLButtonElement)
    expect(clearButton).not.toBeNull()
    await userEvent.click(clearButton as HTMLButtonElement)
    expect(textarea.value).toBe('')
    expect(rendered.asFragment()).toMatchSnapshot()
  })
})
