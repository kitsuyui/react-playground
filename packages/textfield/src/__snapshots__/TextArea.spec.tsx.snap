// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextArea > clear should reset value 1`] = `
<DocumentFragment>
  <textarea>
    Hello, World!
  </textarea>
  <button
    type="button"
  >
    Clear
  </button>
</DocumentFragment>
`;

exports[`TextArea > clear should reset value 2`] = `
<DocumentFragment>
  <textarea>
    Hello, World!2
  </textarea>
  <button
    type="button"
  >
    Clear
  </button>
</DocumentFragment>
`;

exports[`TextArea > clear should reset value 3`] = `
<DocumentFragment>
  <textarea />
  <button
    type="button"
  >
    Clear
  </button>
</DocumentFragment>
`;

exports[`TextArea > should handle input 1`] = `
<DocumentFragment>
  <textarea />
  <button
    type="button"
  >
    Clear
  </button>
</DocumentFragment>
`;

exports[`TextArea > should handle input 2`] = `
<DocumentFragment>
  <textarea>
    Hello, World!
  </textarea>
  <button
    type="button"
  >
    Clear
  </button>
</DocumentFragment>
`;

exports[`TextArea > should handle input with composition change events 1`] = `
<DocumentFragment>
  <textarea />
  <button
    type="button"
  >
    Clear
  </button>
</DocumentFragment>
`;

exports[`TextArea > should render 1`] = `
<DocumentFragment>
  <textarea>
    Hello, World!
  </textarea>
</DocumentFragment>
`;
